### ISSUE TITLE
Hierarchical tags

### ISSUE THREAD

**Issue by Nomadicraty** [laurent22/joplin#375]
Like a Evernote:
![evernote_2018-04-04_10-57-16](https://user-images.githubusercontent.com/32388532/38295545-ffd56944-37f6-11e8-96a5-f55c35f7a47a.png)

---

### Operating system

- [x] Windows
- [x] macOS
- [x] Linux
- [x] Android
- [x] iOS

### Application

- [x] Desktop
- [x] Mobile
- [x] Terminal

**Comment by BenoitBotton** [comment]
yes please, this would help tremendously with using Joplin for https://thesecretweapon.org/

**Comment by nr458h** [comment]
And it would also add the possibility to use it as a kanban board with limitation of wip

**Comment by nr458h** [comment]
Yes you are right it is far from usual board-software.
But with hirarchical tags you could create a tag for each project and group under each project the different stages of wip. 
With the feature request for notes per folder/tag you could see in a second the status of each project ... Not that visual but in my opinion it would do the job

**Comment by cable729** [comment]
To aid in search, I'm mentioning the keywords "nested", "parent", "children".

**Comment by bmscc** [comment]
Dear all,
congratulations for the development of Joplin.

My professional/personal note-taking workflow (Evernote Web), heavily relies on **Hierarchical Tags** to organize information:
example:
![hierarchical_tags](https://user-images.githubusercontent.com/48456938/54878481-f73d5f00-4e2d-11e9-8974-87869a7997b2.jpg)
Having a "flat" Tag visualization instead of a Tree view makes it a lot harder to browse the database.

Are there any plans to implement this feature in a foreseeable future?

Thanks in advance

Best regards

**Comment by JLKM** [comment]
Hierarchical tags would provide a kind of taxonomies. This would be really awesome, and perhaps it's not so hard to implement. How about handling `\\`, `\`, `//`or `/` as separators in tags? Like `country\\region`.

**Comment by jemag** [comment]
I agree that this would be amazing. Having imported my notes from Evernote, instead of having 5-6 top level tags, I now see hundreds of them in the tag list. Definitely making me reconsider if I should switch over or not

**Comment by luckydays** [comment]
Hi@laurent22.
Could you get back to this feature request, please?
We're really struggling without here :)

In general - how much realistic this can be in the nearest future, and how does it currently look like from the WIP features?

Thank you for this amazing product!

**Comment by cspk** [comment]
This would be a really, really nice feature. Moreover, if you have hierarchical tags, you no longer need notebooks at all—your notebooks may be represented as top-level tags. Notebooks in fact start feel too rigid and restricting after you've become acquainted with the hierarchical tags. I've yet failed to find a self-hosted, no BS note-taking software implementing this (Standard Notes has it via paid extension; QOwnNotes gets it out of the box but lacks an Android app; Trilium has similar, but inferior concept of cloning).

**Comment by pconwell** [comment]
I agree with @cspk . Organizing notes into 'notebooks' feels unnecessarily rigid. It makes much more sense to be able to tag a note with multiple tags (versus only being able to put it in one notebook). The topic of a note might overlap with multiple topics, and if you are forced into organizing by notebook you can have a really hard time deciding which notebook to put something in. And when you try to find the note 6 months later and you can't remember what you filed it under it becomes frustrating. Tags, on the other hand, make much more sense.

I mean, gmail ditched folders for tags and it has been hugely successful.

**Comment by cablop** [comment]
I adhere to this request.

It is one of the two features that are preventing me to move my GTD system from Evernote to Joplin. Nested Tags would be so useful.

**Comment by cablop** [comment]
> This would be a really, really nice feature. Moreover, if you have hierarchical tags, you no longer need notebooks at all—your notebooks may be represented as top-level tags. Notebooks in fact start feel too rigid and restricting after you've become acquainted with the hierarchical tags. I've yet failed to find a self-hosted, no BS note-taking software implementing this.
> 
> Standard Notes has it, but I had troubles setting it up, plus it tags notes both with the parent tag and a child tag. This should _not_ be the case: if a note is tagged with a subtag of a tag, it must imply tagging with the parent tag. QOwnNotes gets it out of the box but lacks an Android app. Trilium has similar, but inferior concept of cloning: you have to do stuff manually to expose notes/folders to various parent folders.

I disagree with getting rid of the Notebooks. Notebooks work good as currently intended, and provide a rigid structure to save the notes in. Tags, in the other hand should be for a more flexible way to store data, and combined you have a drawer, a notebook, to hld the content, while tags describe the status of that content inside that container.

**Comment by pconwell** [comment]
> while tags describe the status of that content inside that container.

That's not the way I envision using tags, personally. Tags, to me, are not a status indicator but rather a topic - basically the same way gmail uses tags. EVERYTHING goes into one 'folder' (all mail) and everything else is just a tag. For example, gmail does not have an actual inbox, it's just items that have the 'inbox' tag.

Now, it's an easy fix (as in easily described, not necessarily easily coded), if you want to use notebooks you can still use them. If you don't want to use notebooks, just put everything into one 'master' notebook.

Personally, if nested tags were an option I would not use the notebook feature at all. Everything would get dumped into one master notebook and I would only use tags. Organizing information into discrete notebooks makes no senses to me. An individual note may fit into multiple 'categories' and currently notebooks do not allow you to have the note in multiple notebooks (I guess you could make duplicates technically?). Imagine you had to go to an extensive training program for work. Would you put it in the work notebook? Training notebook? Or what about a note that has code snippits for various work and personal projects? Does that go into the coding folder? Or the work folder? Or the home folder? All of these are easily fixed by forgetting about folders/notebooks and just tag everything.

Again, all that being said - you don't have to actually get rid of notebooks, just don't use them. But honestly, I don't see the point in using notebooks when tags solve a whole bunch of problems and don't introduce any issues. You can functionally use tags exactly like notebooks anyway.

**Comment by mic704b** [comment]
Further to @pconwell's comment, a notebook could instead be a list of notes selected by a tag filter

For example, I might want a "Work" notebook to contain all notes with tag "work" and without the tag "archived".

Or a "Garbage Bin" notebook might list all notes with the tag "deleted"

**Comment by simonsan** [comment]
The nice stuff about tags is: you can use them as you like. I would prefer the way mic704b wants to use them. But for sure you can also use them the way you want pconwell. :-)

**Comment by pconwell** [comment]
If you've ever used RTM, what @mic704b is describing is a Smart List, which I use quite a bit. Very useful.

**Comment by ivangretsky** [comment]
Hey, bot! This is probably one of the most potentially influential issue for Joplin's future, so please leave it as it is)

**Comment by aashish108** [comment]
BAD BOT!!!

**Comment by nr458h** [comment]
really bad bot, I really want to have this feature

**Comment by pconwell** [comment]
@stalebot - the only reason this is "stale" is @laurent22 hasn't chimed in yet.

**Comment by laurent22** [comment]
Done, it's immune from stale bot now.

**Comment by nr458h** [comment]
could this be a feature implemented during GSoC?

**Comment by EL-SHREIF** [comment]
Is this still available as GSoC project?

**Comment by monkeydump** [comment]
Just another voice here wishing for hierarchical tags & the prospect of GTD paradise!

**Comment by sairajsawant** [comment]
I am interested to work on this during GSOC 2020!

**Comment by pheidrias** [comment]
I would very much appreciate hierarchical tagging, too!

**Comment by luckydays** [comment]
@laurent22 could you perhaps comment on whether this feature is in the pipeline, please?
It would so incredibly awesome to have the hierarchical tags in Joplin!

**Comment by laurent22** [comment]
@vsimkus has implemented it in [this PR](https://github.com/laurent22/joplin/pull/2572) and I'm planning to review it soon.

**Comment by jpessolani** [comment]
Nice! Waiting for Hierarchical tags to become a thing before migrating to Joplin

**Comment by Armagedoom** [comment]
Same here @jpessolani . I want to migrate from [Notable](https://github.com/notable/notable), which already does this, into Joplin, but cant go back away from nested tags, they are just mandatory, you guys will see when they are finally implemented here.
The main reason to migrate is that it has become non-open source recently and you guys already have Dropbox sync, which the other one lacks yet and might be behind a paywall in the future.

**Comment by macgrouther** [comment]
First, thanks for creating Joplin! I have already migrated from Evernote and am very happy with the transition. Please add my name to the list of people requesting hierarchical tags.  I use the stacked notebooks and tags together, so I wouldn't want to see the notebooks feature go away.

**Comment by tawfeqfauzi** [comment]
Hi. Any updates on this feature? I would love to start using and export everything to Joplin. This might very well be the only thing holding me back.

**Comment by amsmeier** [comment]
This feature would be very useful.

**Comment by JeffRockatansky** [comment]
This feature is being developed by @vsimkus.

A PR #2572 has been merged but later reverted due to some performance issues

> It has been reverted due to performance issues. The original author will look into it when he has more time.
> _Originally posted by @tessus in https://github.com/laurent22/joplin/issues/2572#issuecomment-733087807_

**Comment by arkid** [comment]
I'm just jumping ship from EverNote like a ton of people right now due to their disastrous version 10.  Found Joplin but no tag hierarchies is one of two or 2 or 3 reasons why I can't go to it.   Please add this and maybe I can check it out again later.

**Comment by BeatLink** [comment]
I really hope this feature will be implemented soon. I'd be willing to pay a bounty for this issue

**Comment by BeatLink** [comment]
In terms of implementation, the easiest way to do so (if not attempted already) would be to have every tag only have a parent ID (which can be null for top level tags) This is all the information needed to generate ancestry and children trees for the tags on the fly. This is preferable to having each tag contain its full parent/ancestry information

**Comment by BeatLink** [comment]
In fact, it seems that notebooks already use the parent ID method for generating hierarchy. Couldnt tags use the same implementation that notebooks uses for hierarchy?

**Comment by kimcosmos** [comment]
Instead of hierarchical tags I use hierarchical notebooks. Then when I need to manually add secondary tags from other notebooks I can. That lets me search by parent or child. With and without hierarchy.  
However... what I need to make this work well is the ability to batch tag every note in a notebook recursively.

**Comment by gongkui** [comment]
Can tags collapse if they have similar start string?

I use many tasks and make the same category with the save starts. 
For example, art category, I have tags like this 
- art.music
- art.painting
- art.photography

If I can collapse them, it will work similar to nested tags.

Thank you

**Comment by darkphase** [comment]
This would be implemented just as splitting tag strings by **user configurable** seperator ( / . : ) and rendering it into a tree view... 
Creating parent - child relations on tags simply against the tagging methodology.

Strict hierarchies are not really found in real world so;
Also disabling notebooks completely on the left menu with a config parameter would be nice...

**Comment by ivangretsky** [comment]
> 
> 
> This would be implemented just as splitting tag strings by **user configurable** seperator ( / . : ) and rendering it into a tree view... Creating parent - child relations on tags simply against the tagging methodology.
> 
> Strict hierarchies are not really found in real world so; Also disabling notebooks completely on the left menu with a config parameter would be nice...

Maybe the tags could be done as you proposed, but displayed ad hierarchical?

**Comment by darkphase** [comment]
Please Check the APP in the link.. Just about how it handles grouping... 
I'm not making any hidden suggestion about any other feature that app may have.

PURPOSE: Handling categories like that is simply allows us put a note into multiple categories... 

My personal opinion about tag cloud is it would be completely useless in that use case. And in any other similar case,  if you're trying to emulate a tree hierarchy via tags.


https://github.com/notable/notable
Regards

**Comment by JackGruber** [comment]
I don't think so.
Before this feature would implemented the issue https://github.com/laurent22/joplin/issues/4251 would have to be fixed, because the last implementation caused even more performance problems (If I remember correctly).

**Comment by tessus** [comment]
> I'm not familiar with backend functions.. Can you tell me what algorithms you're referring to?

I meant the backend meta data and structure that has to be maintained. While a database should usually be in 3NF and make use of foreign keys and constraints, this is not done because of the sync algorithm.

This means that the applications has to make sure that the data inserted and handled in the database is consistent.

Your idea of adding these tags to a table with n:m relation is basically correct and pretty much what the existing PR does. Your initial comment rather suggested to put all 4 tags in one column and record, which would basically do nothing but allow you to show these 4 tags in the UI, but that's all. This would not solve anything to make hierarchical tags work.

Either way, as long as nobody is revisiting the [PR](https://github.com/laurent22/joplin/pull/2572) (which was subsequently reverted) and making the necessary adjustments, nothing is going to happen in that regard.

**Comment by darkphase** [comment]
> Your initial comment rather suggested to put all 4 tags in one column and record, which would basically do nothing but allow you to show these 4 tags in the UI, but that's all. This would not solve anything to make hierarchical tags work.

I didn't propose such thing... I never talked about database structure... At least the intention wasn't that... I'm sorry if I conveyed the idea badly. 

> I meant the backend meta data and structure that has to be maintained. While a database should usually be in 3NF and make use of foreign keys and constraints, this is not done because of the sync algorithm.

3NF or not, I don't think this is directly relevant to the issue... It's a general thing.

> Your idea of adding these tags to a table with n:m relation is basically correct.

I implemented that in many projects with sqlite.
Not just n:m relations, It makes all the difference, if tags are stored in a way I tried to explain above.

I cannot access the implementation about this, but I suspected that it was trying to construct parent-child relations between tags after splitting them via "/". So is the performance issues while querying deep hierarchies... ( just suspicion of course don't get offended please. If I'm wrong please correct me )

> Either way, as long as nobody is revisiting the PR (which was subsequently reverted) and making the necessary adjustments, nothing is going to happen in that regard.

You should say that nobody has time to implement it at the very beginning =)

**Comment by grm9762** [comment]
As someone who only keeps one notebook, but hundreds of tags, the switch from evernote to joplin is unfortunately impossible.

I have an idea: 

Would it be possible to add at least some kind of "headings" to the tags-overview? I mean only a cosmetic update, i.e. these headings are not tags, but should only help to better organize the real tags.

Hope for your support, that would already bring a huge relief.

**Comment by simonsan** [comment]
> Either way, as long as nobody is revisiting the [PR](https://github.com/laurent22/joplin/pull/2572) (which was subsequently reverted) and making the necessary adjustments, nothing is going to happen in that regard.

As this PR is locked (for whatever reason), was there effort made to profile the changes?
What was the/or a pottential bottle neck?

@vsimkus @tessus

**Comment by vsimkus** [comment]
The feature worked well but it introduced significant performance issues to users with thousands of tags. I planned to come back and fix it but life got to me and I never found time to revisit this. I hope someone else will take it up and bring this feature to Joplin.

As far as I remember the main bottleneck was due to note counting for each tag. It is not straightforward as you may think: you need to check for duplicate notes down in a hierarchy in order to count the number of notes for any upper-level tag otherwise you may count the same note multiple times. Though I think the initial version of this PR did not have such issue; it was introduced during the discussion as there were some changes requested by the maintainers---specifically I was asked not to use SQL views for maintaining the counts (although they were significantly more efficient), since they wanted to move away from any SQL-dialect specific code. That being said I think it should be possible to make it work efficiently with some clever tree algorithms. There is some more discussion in https://github.com/laurent22/joplin/pull/3559, where I've outlined some of the considerations that must be made and possible resolutions to this issue, which I hope may be helpful to anyone who takes it up.

**Comment by simonsan** [comment]
> As far as I remember the main bottleneck was due to note counting for each tag. It is not straightforward as you may think: you need to check for duplicate notes down in a hierarchy in order to count the number of notes for any upper-level tag otherwise you may count the same note multiple times.

What I wonder as you say that, why would you need to count notes at all, doesn't each note have a [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier) or maybe even better a [ULID](https://github.com/ulid/spec) that can be used to identify each note? That being said I have not looked deeply into the code base for now, but I would be interested, because maybe it would make sense to introduce some more basic changes to how the data is handled first to then implement some more efficient algorithm/methods for this issue here?

**Comment by kimcosmos** [comment]
Darkphase, you are asking if the nesting tags remained joined for faster indexing? Minimising n:m:m joint foreign key messy lookups. The same reason file paths are stored entire. It would slightly increase the tag length but SQLite does not have any string type sizes. So another DB might be more efficient if tag heirarchy depth was capped at 255 char. TINYTEXT type allowing 85 to 255 chars in UTF-8 is not in SQLite

**Comment by tessus** [comment]
> specifically I was asked not to use SQL views for maintaining the counts (although they were significantly more efficient)

Maybe this is something that Laurent would reconsider, if it makes the access way more performant.

**Comment by darkphase** [comment]
@kimcosmos oh at last some acton here... 
You may use whatever DB you want use of course...
It was just a simple suggestion... 
But why keep real tag hierarchies.. They are just tags... which we can display like hierarchies...

**Comment by BeatLink** [comment]
Perhaps for note count purposes?

**Comment by darkphase** [comment]
@BeatLink 
I think, it should be easier the way I proposed =)
And to calculate sum over the (seemingly) upper branches
--
SELECT count(*) FROM table WHERE tag LIKE 'Books / %'
--
simple...

And this way you can put some notes on the branches (in my opinion as it should be) not being have to put all of them on leaves...

**Comment by BeatLink** [comment]
huh okay. sounds good then! I'd be excited to see this feature

**Comment by ericvaandering** [comment]
I will send $100 to anyone or the group of people that solves this. This has been the sticking point in migration away from Evernote for every app I have looked at: no way to organize my ~1000 tags.

**Comment by penguin9000** [comment]
I'm a big fan of the way [Hydrus](https://github.com/hydrusnetwork/hydrus) handles tags for its image database. Each image, in addition to having normal tags, can have hierarchical tags added. For example, if I set up "painting" to be a child tag of "art". Whenever I add "painting" to an image, it automatically also adds "art", saving me the trouble of adding both each time I want to tag an image. When "painting" is removed, it also removes "art".

My favorite part of all is how easy it is to create a search for some logical AND/OR/NOT combination of tags. I like @miciasto's idea for dynamic notebooks, which would be equivalent to a search page in Hydrus. I would find it very useful to have a to-do system centered around notebooks that contain notes matching specific criteria.

By default, tags are displayed in order of how frequently they occur. Right-clicking a tag opens a context menu that allows the user to quickly add to the current search instead of needing to type in the search bar and modify some kind of formatted search string. When a search is performed, the tag list updates to display only those tags applied to the images in the current search.

I would love it if some of these tagging features came to Joplin. I'm sharing this because I hope that the Joplin devs might be able to take some inspiration from Hydrus's code to help implement this feature.

**Comment by tsilvs** [comment]
In addition to @ericvaandering suggestion, I am ready to pay 100 eur for this feature.

**Comment by lvankampenhout** [comment]
I'm adding in another € 100 for the person (or group) that brings this feature alive

**Comment by ul** [comment]
+$100 to the bounty from me.

**Comment by popovoleksandr** [comment]
Dear @laurent22 and amazing Joplin team!
Any updates on this feature? I feel a lot of Joplin users are waiting for this and I'm wondering what is required to put it into the scope for one of the future releases.

**Comment by auzout** [comment]
Does there any other software features nest tags? I only found evernote does it well. I need a better alternative.

**Comment by ivan-liadniou-epam** [comment]
@Auzout, Obsidian has that.

**Comment by WillRieske** [comment]
> Does there any other software features nest tags?

@Auzout  https://github.com/standardnotes/app

**Comment by Auzout** [comment]
> @Auzout, Obsidian has that.

Yes, Obsidian works perfectly. The only drawback is that it's not open source. However, since it's free for personal use, it's worth considering migrating to it.

**Comment by grm9762** [comment]
> @Auzout, Obsidian has that.

@ivan-liadniou-epam Thanks for the tip with Obsidian :) I have already imported all my notes from Evernote to Obsidian.

Unfortunately, my tag structure is not correct in Obsidian. Only the tags that are linked to notes are available. All tags that are without notes and only serve as tag headings are not in Obsidian.

Have you successfully managed to import your tags?

**Comment by Auzout** [comment]
> > @Auzout, Obsidian has that.
> 
> @ivan-liadniou-epam Thanks for the tip with Obsidian :) I have already imported all my notes from Evernote to Obsidian.
> 
> Unfortunately, my tag structure is not correct in Obsidian. Only the tags that are linked to notes are available. All tags that are without notes and only serve as tag headings are not in Obsidian.
> 
> Have you successfully managed to import your tags?

you need to reorganize your tags by dragging on tags view

**Comment by auzout** [comment]
This plugin enables obisidian nested tags function.
https://github.com/pjeby/tag-wrangler

**Comment by feliponelguapon** [comment]
I think this feature would be great! :)

**Comment by Aleks-Ya** [comment]
Any news?

**Comment by ivangretsky** [comment]
Something just must be preventing the devs to implements this. Or I do not see why one of the most requested features is still not worked on since 2018. Maybe it is the previously taken architectural decisions / conflicts with existing functionality. Would be nice to have some feedback about this from the @laurent22 or maybe someone else in the know.

**Comment by laurent22** [comment]
For information, the full history of this feature is there:

https://github.com/laurent22/joplin/pull/2572

https://github.com/laurent22/joplin/pull/3559
